aten/src/ATen/core/CachingHostAllocator.h

```diff
@@ -220,50 +224,101 @@ struct alignas(hardware_destructive_interference_size) HostStatsStaged {
+  std::deque<std::pair<E_, B_*>> events_;
+};
+
+// The HostBlockPool owned by one HostPrivatePool cannot share blocks
+// with any other HostPrivatePool. This invariant is necessary to
+// implement cuda graph capture correctly, which depends upon virtual
+// addresses staying alive for the duration of a graph's existence.
+template <typename S_, typename E_, typename B_>
+struct HostPrivatePool {
+  explicit HostPrivatePool(c10::MempoolId_t id_) : id(id_) {}
+  HostPrivatePool(const HostPrivatePool&) = delete;
+  HostPrivatePool& operator=(const HostPrivatePool&) = delete;
+
+  c10::MempoolId_t id;
+  int use_count{1};
+  HostBlockPool<S_, E_, B_> blocks;
+};
+
+
 template <
     typename S,
     typename E,
     typename B = HostBlock<S>>
 struct CachingHostAllocatorImpl {
+
+  using BlockPool = HostBlockPool<S, E, B>;
+  using PrivatePool = HostPrivatePool<S, E, B>;
+
   virtual ~CachingHostAllocatorImpl() {
     if (active_) {
       active_ = false;
       getBackgroundThreadPool()->waitWorkComplete();
     }
   }
 
- public:
   // return data_ptr and block pair.
   virtual std::pair<void*, void*> allocate(size_t size) {
     if (size == 0) {
       return {nullptr, nullptr};
     }
 
+    auto&& [mempool_id, pool] = get_allocation_pool(get_current_stream());
+
     // If we are using background threads, we can process events in the
     // background.
     if (!pinned_use_background_threads()) {
-      process_events();
+      process_events(pool);
     }
 
     // Round up the allocation to the nearest power of two to improve reuse.
     // These power of two sizes are also used to index into the free list.
     size_t roundSize = c10::llvm::PowerOf2Ceil(size);
 
-    // First, try to allocate from the free list
-    auto* block = get_free_block(roundSize);
+    // First, try to allocate from the free list of the chosen pool
+    auto* block = get_free_block(roundSize, pool);
     if (block) {
+      block->was_allocated_during_stream_capture_ = stream_is_capturing(get_current_stream());
       return {block->ptr_, reinterpret_cast<void*>(block)};
     }
 
     // Check in the recently freed blocks with pending events to see if we
     // can reuse them. Call get_free_block again after processing events
     if (pinned_use_background_threads()) {
+      process_events_for_specific_size(roundSize, pool);
+      block = get_free_block(roundSize, pool);
+      if (block) {
+        block->was_allocated_during_stream_capture_ = stream_is_capturing(get_current_stream());
+        return {block->ptr_, reinterpret_cast<void*>(block)};
+      }
+
       // Launch the background thread and process events in a loop.
       static bool background_thread_flag [[maybe_unused]] = [this] {
         active_ = true;
         getBackgroundThreadPool()->run([&]() {
           while (active_) {
-            process_events();
+            // Background thread conservatively processes default pool events.
+            process_events(default_pool_);
```

Comment:
I think we are ok with this right? We are not planning on emptying graph blocks, and we are probably fine with just calling empty_cache on them specifically once graph is no longer alive